### PR TITLE
Implement shellSplitString for proper handling of NIX_SSHOPTS with spaces and quotes

### PR DESCRIPTION
**Pull request by elikoga**
Solves #5181 

## Motivation

With NIX_SSHOPTS='-o ProxyCommand="ssh -W %h:%p ..."' nix-copy-closure tool will fail with error.

## Context

See #5181 

Added `shellSplitString` to [strings.cc](https://github.com/NixOS/nix/compare/master...elikoga:nix:ssh-opts-split-as-shell-args?expand=1#diff-a37a841d9f9dcbfe63d651a4a4ee1962b7f000c663bbb83d0e66393fda3895a9)

---

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

File: src/libutil/strings.cc [hunk 1]
```diff
@@ -48,4 +48,69 @@ template std::string dropEmptyInitThenConcatStringsSep(std::string_view, const s
...
+{
+    std::list<std::string> result;
+    std::string current;
+    bool inQuoteSingle = false;
+    bool inQuoteDouble = false;
+    bool escaped = false;
+    bool lastCharWasQuote = false;
+    for (char c : s) {
+        lastCharWasQuote = false;
+        if (escaped) {
```
Comment (edolstra) [inline src/libutil/strings.cc:67]: Maybe this should be disabled while in a single quoted string, since
```
$ echo "\\"
\

$ echo '\\'
\\
```

The logic of this function is quite hard to follow. Maybe it should be something like:
```
auto pop = [&]() -> char {
    auto c = s.front();
    s.remove_prefix(1);
    return s;
}

while (!s.empty()) {
    auto c = pop();
    if (c == '\'') {
        ... copy characters until we encounter another '
    } else if (c == '"') {
        ... copy and unescape characters until we encounter another "
    } else if (c == '\\') {
        if (!s.empty())
            current.push_back(pop());
    } else if (c is whitespace) {
        result.push_back(current);
        current.clear();        
    } else {
        current.push_back(c);
    }
}
```
or something along those lines.

Comment (elikoga) [reply src/libutil/strings.cc]: The current implementation has been adjusted to look like this

**Review by Mic92 — COMMENTED** [review]
Since this is a behavior change. Can we get release notes on this?

**Comment by Mic92** [comment]
Indeed this would be nice to have for nixos-anywhere.

**Comment by elikoga** [comment]
I'll get to it today. Do we want any more changes before we include this? Make the function generic or move it to the usage site?

File: src/libutil/strings.cc [hunk 2]
```diff
@@ -48,4 +48,107 @@ template std::string dropEmptyInitThenConcatStringsSep(std::string_view, const s
...
+            } else if (c == '\\') {
+                escaping = true;
+            } else if (c == '"') {
+                return;
+            } else {
+                pushChar(c);
+            }
+        }
+        if (s.empty()) {
+            throw std::runtime_error("unterminated double quote (probably in NIX_SSHOPTS)");
```
Comment (Mic92) [inline src/libutil/strings.cc:114]:
```suggestion
            throw Error("unterminated double quote");
```

File: src/libutil/strings.cc [hunk 3]
```diff
@@ -48,4 +48,107 @@ template std::string dropEmptyInitThenConcatStringsSep(std::string_view, const s
...
+        startedCurrent = true;
+        while (!s.empty()) {
+            auto c = pop();
+            if (c == '\'') {
+                return;
+            }
+            pushChar(c);
+        }
+        if (s.empty()) {
+            throw std::runtime_error("unterminated single quote (probably in NIX_SSHOPTS)");
```
Comment (Mic92) [inline src/libutil/strings.cc:128]:
```suggestion
            throw Error("unterminated single quote");
```

File: src/libstore/ssh.cc [hunk 4]
```diff
@@ -41,7 +41,7 @@ void SSHMaster::addCommonSSHOpts(Strings & args)
 {
     auto state(state_.lock());
```
Comment (Mic92) [inline src/libstore/ssh.cc:43]: Unsure if we should actually print the value of NIX_SSHOPTS here, can it contain secrets? But anyway the point is using `addTrace`.

```suggestion
  std::string env = getEnv("NIX_SSHOPTS").value_or("");
  try {
    std::list<std::string> parts = shellSplitString(env);
  } catch (Error & e) { 
     e.addTrace({}, "while parsing NIX_SSHOPTS '%s'", env);
     throw; 
 } 
```

**Comment by elikoga** [comment]
I am not sure how to interpret the current CI failures

**Comment by Mic92** [comment]
It's the test-suite but with a different nix-daemon: 
https://github.com/NixOS/nix/blob/f1187cb696584739884687d788a6fbb4dd36c61c/packaging/hydra.nix#L24

**Comment by Mic92** [comment]
I re-triggered the test run to see if the issue persists.

**Review by Mic92 — APPROVED** [review]
Haven't go around testing it, yet. But looks good to me.

**Comment by Mic92** [comment]
@mergify queue

**Comment by Mic92** [comment]
@mergify rebase
